api/src/services/authorization.ts

```diff
@@ -606,9 +606,14 @@ export class AuthorizationService {
 			const result = await itemsService.readMany(pk, { ...query, limit: pk.length }, { permissionsAction: action });
 			if (!result) throw new ForbiddenError();
 			if (result.length !== pk.length) throw new ForbiddenError();
-		} else {
+		} else if (pk) {
 			const result = await itemsService.readOne(pk, query, { permissionsAction: action });
 			if (!result) throw new ForbiddenError();
+		} else {
+			query.limit = 1;
+			const result = await itemsService.readByQuery(query, { permissionsAction: action });
+			const access = (action === 'create' && result) || result[0];
+			if (!access) throw new ForbiddenError();
```

Comment:
What is the reasoning here for checking access like this? Is `result.length > 0` not enough?

On a side note `readByQuery` seems to always return an array. Empty arrays are truthy so we should specifically check the length.

Reply:
For the `create` action we expect an empty array (because the singleton doesn't exist yet), while for the other actions `result[0]` is expected to defined.

The logic follows the other check above:
https://github.com/directus/directus/blob/2a2dc5fcd21057dd611c1594cbb39ddabb417d42/api/src/services/authorization.ts#L606-L608

We could however use something like this to be more explicit:

```js
const access = action === 'create' ? result.length === 0 : !!result[0];
if (!access) throw new ForbiddenError();
```

But still feels safer to do a `!result` check beforehand, in case `result` should not be defined for some reason (e.g. hooks 🤔)
```
if (!result) throw new ForbiddenError();
const access = action === 'create' ? result.length === 0 : !!result[0];
if (!access) throw new ForbiddenError();
```

Reply:
Updated 👍